README.md
# Autohook 🎣 + Fastify 

This is a fork of [Autohook](https://github.com/twitterdev/autohook) that uses [Fastify](https://www.fastify.io/) instead of needle.

![Demo](https://github.com/twitterdev/autohook/raw/master/demo.gif)

* 🚀 Spawns a server for you
* ⚙️ Registers a webhook (it removes existing webhooks if you want, and you can add more than one webhook if your Premium subscription supports it)
* ✅ Performs the CRC validation when needed
* 📝 Subscribes to your current user's context (you can always subscribe more users if you need)
* 🎧 Exposes a listener so you can pick up Account Activity events and process the ones you care about

## Usage

You can use Autohook as a module or as a command-line tool.

### Node.js module

```js
const { Autohook } = require('twitter-autohook');

(async ƛ => {
  const webhook = new Autohook();
  
  // Removes existing webhooks
  await webhook.removeWebhooks();
  
  // Listens to incoming activity
  webhook.on('event', event => console.log('Something happened:', event));
  
  // Starts a server and adds a new webhook
  await webhook.start();
  
  // Subscribes to a user's activity
  await webhook.subscribe({oauth_token, oauth_token_secret});
})();
```

### Command line

Starting Autohook from the command line is useful when you need to test your connection and subscriptions.

When started from the command line, Autohook simply provisions a webhook, subscribes your user (unless you specify `--do-not-subscribe-me`), and echoes incoming events to `stdout`.

```bash
# Starts a server, removes any existing webhook, adds a new webhook, and subscribes to the authenticating user's activity.
$ autohook -rs

# All the options
$ autohook --help
```

## OAuth

Autohook works only when you pass your OAuth credentials. You won't have to figure out OAuth by yourself – Autohook will work that out for you.

You can pass your OAuth credentials in a bunch of ways.

## Dotenv (~/.env.twitter)

Create a file named `~/.env.twitter` (sits in your home dir) with the following variables:

```bash
TWITTER_CONSUMER_KEY= # https://developer.twitter.com/en/apps ➡️ Your app ID ➡️ Details ➡️ API key
TWITTER_CONSUMER_SECRET= # https://developer.twitter.com/en/apps ➡️ Your app ID ➡️ Details ➡️ API secret key
TWITTER_ACCESS_TOKEN= # https://developer.twitter.com/en/apps ➡️ Your app ID ➡️ Details ➡️ Access token
TWITTER_ACCESS_TOKEN_SECRET= # https://developer.twitter.com/en/apps ➡️ Your app ID ➡️ Details ➡️ Access token secret
TWITTER_WEBHOOK_ENV= # https://developer.twitter.com/en/account/environments ➡️ One of 'Dev environment label' or 'Prod environment label'
NGROK_AUTH_TOKEN= # https://ngrok.com/ - (optional) Create a free account to get your auth token for stable tunnels
```

Autohook will pick up these details automatically, so you won't have to specify anything in code or via CLI.

## Env variables

Useful when you're deploying to remote servers, and can be used in conjunction with your dotenv file.

```bash

# To your current environment
export TWITTER_CONSUMER_KEY= # https://developer.twitter.com/en/apps ➡️ Your app ID ➡️ Details ➡️ API key
export TWITTER_CONSUMER_SECRET= # https://developer.twitter.com/en/apps ➡️ Your app ID ➡️ Details ➡️ API secret key
export TWITTER_ACCESS_TOKEN= # https://developer.twitter.com/en/apps ➡️ Your app ID ➡️ Details ➡️ Access token
export TWITTER_ACCESS_TOKEN_SECRET= # https://developer.twitter.com/en/apps ➡️ Your app ID ➡️ Details ➡️ Access token secret
export TWITTER_WEBHOOK_ENV= # https://developer.twitter.com/en/account/environments ➡️ One of 'Dev environment label' or 'Prod environment label'
export NGROK_AUTH_TOKEN= # https://ngrok.com/ - (optional) Create a free account to get your auth token for stable tunnels

# To other services, e.g. Heroku
heroku config:set TWITTER_CONSUMER_KEY=value TWITTER_CONSUMER_SECRET=value TWITTER_ACCESS_TOKEN=value TWITTER_ACCESS_TOKEN_SECRET=value TWITTER_WEBHOOK_ENV=value NGROK_AUTH_TOKEN=value
```
## Directly

Not recommended, because you should always [secure your credentials](https://developer.twitter.com/en/docs/basics/authentication/guides/securing-keys-and-tokens.html).

### Node.js

```js
new Autohook({
  token: 'value',
  token_secret: 'value',
  consumer_key: 'value',
  consumer_secret: 'value',
  ngrok_secret: 'value', // optional
  env: 'env',
  port: 1337
});
```

### CLI

```bash
$ autohook \
  --token $TWITTER_ACCESS_TOKEN \
  --secret $TWITTER_ACCESS_TOKEN_SECRET \
  --consumer-key $TWITTER_CONSUMER_KEY \
  --consumer-secret $TWITTER_CONSUMER_SECRET \
  --env $TWITTER_WEBHOOK_ENV \
  --ngrok-secret $NGROK_AUTH_TOKEN # optional
```

## Install

```bash
# npm
$ npm i -g twitter-autohook

# Yarn
$ yarn global add twitter-autohook
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.0.1](https://github.com/twitterdev/autohook/compare/v2.0.0...v2.0.1) (2022-03-11)


### Bug Fixes

* ignore .env file ([eee5208](https://github.com/twitterdev/autohook/commit/eee52080acf6d1746a5313862114e28f323d194d))

## [2.0.0](https://github.com/twitterdev/autohook/compare/v1.7.1...v2.0.0) (2022-03-10)


### ⚠ BREAKING CHANGES

* add custom twitter host url

### Features

* add custom twitter host url ([f99ded8](https://github.com/twitterdev/autohook/commit/f99ded85865ebfcaf0d30702cf4246929a15e146))
* add fastify as a server startup option ([5a5cfa8](https://github.com/twitterdev/autohook/commit/5a5cfa83e55571c6b8103f392823568104d0b637))


### Bug Fixes

* add console logs ([b2bf087](https://github.com/twitterdev/autohook/commit/b2bf0876377cce84fcfda4d59286319c9e04f262))
* add custom webhook routing ([f73b4fd](https://github.com/twitterdev/autohook/commit/f73b4fd558fff09a988c3c5200194225c88bcda5))
* add http back ([2883717](https://github.com/twitterdev/autohook/commit/28837171aa39fc823d6d6da34452f3d162120118))
* add Procfle ([27cb786](https://github.com/twitterdev/autohook/commit/27cb786c2f69ba864145d8b22048b68ee33740d9))
* remove requirement for webhookURL ([63ff1bb](https://github.com/twitterdev/autohook/commit/63ff1bb35a109979c870aeb1913ef38a43e862e7))
* simplify ([099b2ed](https://github.com/twitterdev/autohook/commit/099b2ed5917f271e45e11ff2ffd93f1ee1839f3d))
* wip ([3dd024c](https://github.com/twitterdev/autohook/commit/3dd024c7f115e51e48869e1ebed46b829736abd0))


test.js
const {Autohook} = require('./index');

const qs = require('querystring');
const request = require('request');
const readline = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout
});
const util = require('util');
const path = require('path');
const os = require('os');

const get = util.promisify(request.get);
const post = util.promisify(request.post);
const sleep = util.promisify(setTimeout);

const requestTokenURL = new URL('https://api.twitter.com/oauth/request_token');
const accessTokenURL = new URL('https://api.twitter.com/oauth/access_token');
const authorizeURL = new URL('https://api.twitter.com/oauth/authorize');

async function input(prompt) {
  return new Promise(async (resolve, reject) => {
    readline.question(prompt, (out) => {
      readline.close();
      resolve(out);
    });
  });
}

async function accessToken({oauth_token, oauth_token_secret}, verifier) {
  const oAuthConfig = {
    consumer_key: process.env.TWITTER_CONSUMER_KEY,
    consumer_secret: process.env.TWITTER_CONSUMER_SECRET,
    token: oauth_token,
    token_secret: oauth_token_secret,
    verifier: verifier,
  }; 
  
  const req = await post({url: accessTokenURL, oauth: oAuthConfig});
  if (req.body) {
    return qs.parse(req.body);
  } else {
    throw new Error('Cannot get an OAuth access token');
  }
}

async function requestToken() {
  const oAuthConfig = {
    callback: 'oob',
    consumer_key: process.env.TWITTER_CONSUMER_KEY,
    consumer_secret: process.env.TWITTER_CONSUMER_SECRET,
  };

  const req = await post({url: requestTokenURL, oauth: oAuthConfig});
  if (req.body) {
    return qs.parse(req.body);
  } else {
    throw new Error('Cannot get an OAuth request token');
  }
}

async function markAsRead(messageId, senderId, auth) {
  const requestConfig = {
    url: 'https://api.twitter.com/1.1/direct_messages/mark_read.json',
    form: {
      last_read_event_id: messageId,
      recipient_id: senderId,
    },
    oauth: auth,
  };

  await post(requestConfig);
}

async function indicateTyping(senderId, auth) {
  const requestConfig = {
    url: 'https://api.twitter.com/1.1/direct_messages/indicate_typing.json',
    form: {
      recipient_id: senderId,
    },
    oauth: auth,
  };

  await post(requestConfig);
}

async function sayHi(event, oauth) {
  // Only react to direct messages
  if (!event.direct_message_events) {
    return;
  }

  const message = event.direct_message_events.shift();
 
  // Filter out messages created by the the authenticating users (to avoid sending messages to oneself)
  if (message.message_create.sender_id === oauth.user_id) {
    return;
  }

  const oAuthConfig = {
    token: oauth.oauth_token,
    token_secret: oauth.oauth_token_secret,
    consumer_key: oauth.consumer_key,
    consumer_secret: oauth.consumer_secret,
  };


  await markAsRead(message.message_create.id, message.message_create.sender_id, oAuthConfig);
  await indicateTyping(message.message_create.sender_id, oAuthConfig);
  const senderScreenName = event.users[message.message_create.sender_id].screen_name;

  console.log(`${senderScreenName} says ${message.message_create.message_data.text}`);

  const requestConfig = {
    url: 'https://api.twitter.com/1.1/direct_messages/events/new.json',
    oauth: oAuthConfig,
    json: {
      event: {
        type: 'message_create',
        message_create: {
          target: {
            recipient_id: message.message_create.sender_id,
          },
          message_data: {
            text: `Hi @${senderScreenName}! 👋`,
          },
        },
      },
    },
  };
  await post(requestConfig);
}

(async () => {
  try {

    // Get request token
    const oAuthRequestToken = await requestToken();
    
    // Get authorization
    authorizeURL.searchParams.append('oauth_token', oAuthRequestToken.oauth_token);
    console.log('Please go here and authorize:', authorizeURL.href);
    const pin = await input('Paste the PIN here: ');
    
    // Get the access token
    const userToMonitor = await accessToken(oAuthRequestToken, pin.trim());
    const webhook = new Autohook({
      token: process.env.TWITTER_ACCESS_TOKEN,
      token_secret: process.env.TWITTER_ACCESS_TOKEN_SECRET,
      consumer_key: process.env.TWITTER_CONSUMER_KEY,
      consumer_secret: process.env.TWITTER_CONSUMER_SECRET,
      env: process.env.TWITTER_WEBHOOK_ENV});

    webhook.on('event', async (event) => {
      await sayHi(event, {
        oauth_token: userToMonitor.oauth_token,
        oauth_token_secret: userToMonitor.oauth_token_secret,
        user_id: userToMonitor.user_id,
        consumer_key: process.env.TWITTER_CONSUMER_KEY,
        consumer_secret: process.env.TWITTER_CONSUMER_SECRET,
        reset: true,
      });
    });

    await webhook.removeWebhooks();
    await webhook.start();
    await webhook.subscribe(userToMonitor);
    
  } catch(e) {
    console.error(e);
    process.exit(-1);
  }
})();

test/oauth.js
const assert = require('assert');
const oauthInstance = require('../oauth');
const URL = require('url').URL;

const oAuthConfig = {
  oauth: {
    consumer_key: 'consumer_key',
    consymer_secret: 'consumer_secret',
    token: 'test_user_token',
    token_secret: 'test_user_token_secret',
  },
};

const signatures = {
  baseUrl: 'OAuth oauth_consumer_key="consumer_key", oauth_nonce="GXjhffMbAMz2qblDzzgYbP4ZkfPp7RGmhry5Upatw", oauth_signature="RGpMnOI1WxKDZtgORbi32uc8P%2BY%3D", oauth_signature_method="HMAC-SHA1", oauth_timestamp="1583967563", oauth_token="test_user_token", oauth_version="1.0"',
  urlWithParams: 'OAuth oauth_consumer_key="consumer_key", oauth_nonce="xaCqEY8Ed9vnfFvZJsu8AjSF1", oauth_signature="LYT0mRAq62MXsnxsTOppEtViTUs%3D", oauth_signature_method="HMAC-SHA1", oauth_timestamp="1583967750", oauth_token="test_user_token", oauth_version="1.0"',
};

const baseUrl = new URL('https://api.twitter.com/1.1/account/verify_credentials.json')
oauthInstance.setNonceFn(() => 'GXjhffMbAMz2qblDzzgYbP4ZkfPp7RGmhry5Upatw');
oauthInstance.setTimestampFn(() => '1583967563');
assert.throws(() => {
  // non-object and non string body throws TypeError
  oauthInstance.oauth(baseUrl, 'GET', oAuthConfig, () => {});
}, {
  name: 'TypeError',
});
assert.equal(oauthInstance.oauth(baseUrl, 'GET', oAuthConfig), signatures.baseUrl);
assert.equal(oauthInstance.oauth(baseUrl, 'GET', oAuthConfig, {}), signatures.baseUrl);
assert.equal(oauthInstance.oauth(baseUrl, 'GET', oAuthConfig, ''), signatures.baseUrl);

const urlWithParams = new URL('https://api.twitter.com/1.1/account/verify_credentials.json')
urlWithParams.searchParams.append('param_test', '1');
urlWithParams.searchParams.append('example', '1');
oauthInstance.setNonceFn(() => 'xaCqEY8Ed9vnfFvZJsu8AjSF1');
oauthInstance.setTimestampFn(() => '1583967750');
assert.equal(oauthInstance.oauth(urlWithParams, 'GET', oAuthConfig), signatures.urlWithParams);

const urlWithBodyParams = new URL('https://api.twitter.com/1.1/account/verify_credentials.json');
const bodyParams = {
  param_test: '1',
  example: '1'
};

assert.equal(oauthInstance.oauth(urlWithParams, 'GET', oAuthConfig, bodyParams), signatures.urlWithParams);
const bodyParamsString = 'param_test=1&example=1';
assert.equal(oauthInstance.oauth(urlWithParams, 'GET', oAuthConfig, bodyParamsString), signatures.urlWithParams);

test/client.js
const { post } = require('../client');
const oauthInstance = require('../oauth');
const qs = require('querystring');
const nock = require('nock');
const assert = require('assert');

(async () => {
  const oAuthConfig = {
    consumer_key: 'consumer_key',
    consumer_secret: 'consumer_secret',
    token: 'test_user_token',
    token_secret: 'test_user_token_secret',
  };

  const baseURL = 'https://example.com';
  const formRoute = '/form';
  const jsonRoute = '/json';
  const oauthRoute = '/oauth';

  const mockData = {
    form: {
      nonce: 'v9jKGGxfQiD2u09dgcFucmyCXKckcpGprlojZhxV4',
      timestamp: '1584380758',
      body: qs.stringify({test_value: 42, form: true, json: false}),
      contentType: 'application/x-www-form-urlencoded',
      oauth: 'OAuth oauth_consumer_key="consumer_key", oauth_nonce="v9jKGGxfQiD2u09dgcFucmyCXKckcpGprlojZhxV4", oauth_signature="7KNZQnQi9ZnV9289OmqE4Xg7dck%3D", oauth_signature_method="HMAC-SHA1", oauth_timestamp="1584380758", oauth_token="test_user_token", oauth_version="1.0"',  
    },
    json: {
      nonce: 'kCMhnYwha4MIuIIRVbYt62A9eoIboIP8',
      timestamp: '1584407161',
      body: {test_value: 42, form: false, json: true},
      contentType: 'application/json; charset=utf-8',
      oauth: 'OAuth oauth_consumer_key="consumer_key", oauth_nonce="kCMhnYwha4MIuIIRVbYt62A9eoIboIP8", oauth_signature="4oV4jRsCON%2FbLeCePh5zR%2Bd73MI%3D", oauth_signature_method="HMAC-SHA1", oauth_timestamp="1584407161", oauth_token="test_user_token", oauth_version="1.0"',  
    },
  };

  const replyCallback = function(route, requestBody) {
    return [200, {
      body: requestBody,
      headers: {
        contentType: this.req.headers['content-type'],
        oauth: this.req.headers.authorization || null,  
      },
    }];
  };

  const scope = nock(baseURL)
    .post(formRoute).reply(replyCallback)
    .post(jsonRoute).reply(replyCallback)
    .post(`${formRoute}${oauthRoute}`).reply(replyCallback)
    .post(`${jsonRoute}${oauthRoute}`).reply(replyCallback);

  const formResponse = await post({url: `${baseURL}${formRoute}`, body: mockData.form.body});
  assert.deepEqual(formResponse.body.body, mockData.form.body);
  assert.equal(formResponse.body.headers.contentType, mockData.form.contentType);
  
  oauthInstance.setNonceFn(() => mockData.form.nonce);
  oauthInstance.setTimestampFn(() => mockData.form.timestamp);
  const formOAuthResponse = await post({url: `${baseURL}${formRoute}${oauthRoute}`, body: mockData.form.body, options: {oauth: oAuthConfig}});
  assert.deepEqual(formOAuthResponse.body.body, mockData.form.body);
  assert.equal(formOAuthResponse.body.headers.contentType, mockData.form.contentType);
  assert.equal(formOAuthResponse.body.headers.oauth, mockData.form.oauth);

  const jsonResponse = await post({url: `${baseURL}${jsonRoute}`, body: mockData.json.body, options: {json: true}});
  assert.deepEqual(jsonResponse.body.body, mockData.json.body);
  assert.equal(jsonResponse.body.headers.contentType, mockData.json.contentType);

  oauthInstance.setNonceFn(() => mockData.json.nonce);
  oauthInstance.setTimestampFn(() => mockData.json.timestamp);
  const jsonOAuthResponse = await post({url: `${baseURL}${jsonRoute}${oauthRoute}`, body: mockData.json.body, options: {oauth: oAuthConfig, json: true}});
  assert.deepEqual(jsonOAuthResponse.body.body, mockData.json.body);
  assert.equal(jsonOAuthResponse.body.headers.contentType, mockData.json.contentType);
  assert.equal(jsonOAuthResponse.body.headers.oauth, mockData.json.oauth);

  nock.cleanAll();
})();

cli.js
#!/usr/bin/env node
const {
  Autohook,
  TooManyWebhooksError,
  UserSubscriptionError,
} = require('.');
const path = require('path');
const os = require('os');

require('dotenv').config({path: path.resolve(os.homedir(), '.env.twitter')});


const argv = require('commander')
  .description('Zero configuration setup Twitter Account Activity API webhooks (Premium).\n\nAll parameters are optional if the corresponding env variable is defined in your env file or in ~/.env.twitter.')
  .option('--token <token>', 'your OAuth access token. (Env var: TWITTER_ACCESS_TOKEN)')
  .option('--secret <secret>', 'your OAuth access token secret. (Env var: TWITTER_ACCESS_TOKEN_SECRET)')
  .option('--consumer-key <consumerKey>', 'your OAuth consumer key. (Env var: TWITTER_CONSUMER_KEY)')
  .option('--consumer-secret <consumerSecret>', 'your OAuth consumer secret. (Env var: TWITTER_CONSUMER_SECRET)')
  .option('--ngrok-secret <authToken>', 'your ngrok authtoken. (Env var: NGROK_AUTH_TOKEN)')
  .option('--env <env>', 'your Premium environment label as defined in https://developer.twitter.com/en/account/environments. (Env var: TWITTER_WEBHOOK_ENV)')
  .option('--port <port>', 'port where the local HTTP server should run. Default: 1337. (Env var: PORT)')
  .option('--url <url>', 'URL to an existing webhook configured to respond to Twitter')
  .option('-s, --subscribe-me', 'subscribes the app to listen to activities from your user context specified by the current access token credentials')
  .option('--subscribe <accessToken:accessTokenSecret>', 'subscribes to activities of the Twitter user idenfified by the specified OAuth credentials', (val, prev) => {
    const [oauth_token, oauth_secret] = val.split(':');
    const oauth = {oauth_token, oauth_secret};
    prev.push(oauth);
    return prev;
  }, [])
  .option('-r, --reset', 'remove existing webhooks from the specified environment before starting a new instance')
  .parse(process.argv);

const webhook = new Autohook({
  token: argv.token || process.env.TWITTER_ACCESS_TOKEN,
  token_secret: argv.secret || process.env.TWITTER_ACCESS_TOKEN_SECRET,
  consumer_key: argv.consumerKey || process.env.TWITTER_CONSUMER_KEY,
  consumer_secret: argv.consumerSecret || process.env.TWITTER_CONSUMER_SECRET,
  env: argv.env || process.env.TWITTER_WEBHOOK_ENV,
  port: argv.port || process.env.PORT,
});

webhook.on('event', event => console.log(JSON.stringify(event, null, 2)));

const subscribe = async (auth) => {
  try {
    webhook.subscribe(auth);
  } catch(e) {
    console.error(e.message);
    process.exit(-1);
  }
  
}

(async () => {
  if (!!argv.reset) {
    await webhook.removeWebhooks();
  }

  try {
    await webhook.start(argv.url || null);  
  } catch(e) {
    switch (e.constructor) {
      case TooManyWebhooksError:
        console.error('Cannot add webhook: you have exceeded the number of webhooks available', 
          `to you for the '${argv.env || process.env.TWITTER_WEBHOOK_ENV}' environment.`,
          `Use 'autohook -r' to remove your existing webhooks or remove callbacks manually`,
          'using the Twitter API.');
        break;
      default:
        console.error('Error:', e.message); 
        break;        
    }

    process.exit(-1);

  }

  if (!!argv.subscribeMe) {
    await subscribe({
      oauth_token: argv.token || process.env.TWITTER_ACCESS_TOKEN,
      oauth_token_secret: argv.secret || process.env.TWITTER_ACCESS_TOKEN_SECRET,
    });    
  }

  for (oauth in argv.subscribe) {
    await subscribe(oauth);
  }

})();


index.js
#!/usr/bin/env node
/* global process, Buffer */
const ngrok = require('ngrok');
const http = require('http');
const url = require('url');
const crypto = require('crypto');
const path = require('path');
const os = require('os');
const EventEmitter = require('events');
const URL = require('url').URL;
const bearerToken = require('./bearer-token');
const { get, post, del } = require('./client');
const Fastify = require('fastify');
const FastifyRawBody = require('fastify-raw-body');

const {
  TooManySubscriptionsError,
  UserSubscriptionError,
  WebhookURIError,
  tryError,
} = require('./errors');

require('dotenv').config({path: path.resolve(os.homedir(), '.env.twitter')});

const DEFAULT_PORT = 1337;
const WEBHOOK_ROUTE = '/webhook';

let _getSubscriptionsCount = null;
const getSubscriptionsCount = async (auth) => {
  if (_getSubscriptionsCount) {
    return _getSubscriptionsCount;
  }

  const token = await bearerToken(auth);
  const requestConfig = {
    url: 'https://api.twitter.com/1.1/account_activity/all/subscriptions/count.json',
    options: { 
      bearer: token 
    },
  };

  const response = await get(requestConfig);

  const error = tryError(response);
  if (error) {
    throw error;
  }

  _getSubscriptionsCount = response.body;
  return _getSubscriptionsCount;
};

const updateSubscriptionCount = (increment) => {
  if (!_getSubscriptionsCount) {
    return;
  }

  _getSubscriptionsCount.subscriptions_count += increment;
};

const deleteWebhooks = async (webhooks, auth, env) => {
  console.log('Removing webhooks…');
  for (const {id, url} of webhooks) {
    const requestConfig = {
      url: `https://api.twitter.com/1.1/account_activity/all/${env}/webhooks/${id}.json`,
      options: {
        oauth: auth,      
      },
    };

    console.log(`Removing ${url}…`);
    await del(requestConfig);
  }
};

const validateWebhook = (token, auth) => {
  const responseToken = crypto.createHmac('sha256', auth.consumer_secret).update(token).digest('base64');
  return {response_token: `sha256=${responseToken}`};
};

const validateSignature = (header, auth, body) => {
  const signatureHeaderName = 'x-twitter-webhooks-signature';

  if (typeof header[signatureHeaderName] === 'undefined') {
    throw new TypeError(`validateSignature: header ${signatureHeaderName} not found`);
  }

  const signature = 'sha256=' + crypto
        .createHmac('sha256', auth.consumer_secret)
        .update(body)
        .digest('base64');

  return crypto.timingSafeEqual(
    Buffer.from(header[signatureHeaderName]),
    Buffer.from(signature));
};

const verifyCredentials = async (auth) => {
  const requestConfig = {
    url: 'https://api.twitter.com/1.1/account/verify_credentials.json',
    options: {
      oauth: auth,
    },
  };

  const response = await get(requestConfig);
  const error = tryError(
    response,
    (response) => new UserSubscriptionError(response));

    if (error) {
    throw error;
  }

  return response.body.screen_name;
};

class Autohook extends EventEmitter {
  constructor({
    token = (process.env.TWITTER_ACCESS_TOKEN || '').trim(),
    token_secret = (process.env.TWITTER_ACCESS_TOKEN_SECRET || '').trim(),
    consumer_key = (process.env.TWITTER_CONSUMER_KEY || '').trim(),
    consumer_secret = (process.env.TWITTER_CONSUMER_SECRET || '').trim(),
    ngrok_secret = (process.env.NGROK_AUTH_TOKEN || '').trim(),
    env = (process.env.TWITTER_WEBHOOK_ENV || '').trim(),
    webhookPath = (process.env.TWITTER_WEBHOOK_PATH || '').trim(),
    port = process.env.PORT || DEFAULT_PORT,
    headers = [],
  } = {}) {

    Object.entries({token, token_secret, consumer_key, consumer_secret, env, port}).map((el) => {
      const [key, value] = el;
      if (!value) {
        throw new TypeError(`'${key}' is empty or not set. Check your configuration and try again.`);
      }
    });
    
    if (!webhookPath) {
      webhookPath = '/';
    }

    super();
    this.auth = {token, token_secret, consumer_key, consumer_secret};
    this.ngrokSecret = ngrok_secret;
    this.env = env;
    this.webhookPath = webhookPath;
    this.port = port;
    this.headers = headers;
  }

  /**
   * Set up the GET and POST endpoints for twitter CRC using Fastify
   * @returns {Promise<string>} fastify.listen promise
   */
  startFastifyServer() {
    const fastify = new Fastify({
      logger: process.env.FASTIFY_LOGGER_FLAG,
    });
    fastify.register(FastifyRawBody, {
      field: 'rawBody',
      global: false,
      encoding: false,
      runFirst: true,
      routes: []
    });
    fastify.get(this.webhookPath, (request, reply) => {
      if (request.query.crc_token) {
        try {
          if (!validateSignature(request.headers, this.auth, url.parse(request.url).query)) {
            console.error('Cannot validate webhook signature');
            return;
          }
        } catch (e) {
          console.error(e);
        }
        const crc = validateWebhook(request.query.crc_token, this.auth);
        reply.statusCode = 200;
        reply.header('content-type', 'application/json');
        reply.send(JSON.stringify(crc));
      }
    });
    
    fastify.post(this.webhookPath, { config: { rawBody: true }}, (request, reply) => {
      try {
        if (!validateSignature(request.headers, this.auth, request.rawBody)) {
          console.error('Cannot validate webhook signature');
          return;
        }
      } catch (e) {
        console.error(e);
      }
      this.emit('event', request, reply);
      reply.statusCode = 200;
      reply.send();
    });
    
    this.server = fastify.server;
    return fastify.listen({
      port: this.port
    });
  }

  startServer() {
    this.server = http.createServer((req, res) => {
      const route = url.parse(req.url, true);

      if (!route.pathname) {
        return;
      }

      if (route.query.crc_token) {
        try {
          if (!validateSignature(req.headers, this.auth, url.parse(req.url).query)) {
            console.error('Cannot validate webhook signature');
            return;
          }
        } catch (e) {
          console.error(e);
        }
        const crc = validateWebhook(route.query.crc_token, this.auth);
        res.writeHead(200, {'content-type': 'application/json'});
        res.end(JSON.stringify(crc));
      }

      if (req.method === 'POST' && req.headers['content-type'] === 'application/json') {
        let body = '';
        req.on('data', (chunk) => {
          body += chunk.toString();
        });
        req.on('end', () => {
          try {
            if (!validateSignature(req.headers, this.auth, body)) {
              console.error('Cannot validate webhook signature');
              return;
            }
          } catch (e) {
            console.error(e);
          }
          this.emit('event', JSON.parse(body), req);
          res.writeHead(200);
          res.end();
        });
      }
    }).listen(this.port);
  }

  async setWebhook(webhookUrl) {
    const parsedUrl = url.parse(webhookUrl);
    if (parsedUrl.protocol === null || parsedUrl.host === 'null') {
      throw new TypeError(`${webhookUrl} is not a valid URL. Please provide a valid URL and try again.`);
    } else if (parsedUrl.protocol !== 'https:') {
      throw new TypeError(`${webhookUrl} is not a valid URL. Your webhook must be HTTPS.`);
    }
  
    console.log(`Registering ${webhookUrl} as a new webhook…`);
    const endpoint = new URL(`https://api.twitter.com/1.1/account_activity/all/${this.env}/webhooks.json`);
    endpoint.searchParams.append('url', webhookUrl);
  
    const requestConfig = {
      url: endpoint.toString(),
      options: {
        oauth: this.auth,
      },
    };
  
    const response = await post(requestConfig);
  
    const error = tryError(
      response,
      (response) => new URIError(response, [
        `Cannot get webhooks. Please check that '${this.env}' is a valid environment defined in your`,
        'Developer dashboard at https://developer.twitter.com/en/account/environments, and that',
        `your OAuth credentials are valid and can access '${this.env}'. (HTTP status: ${response.statusCode})`].join(' '))
    );
  
    if (error) {
      throw error;
    }
    
    return response.body;
  }

  async getWebhooks() {
    console.log('Getting webhooks…');
  
    let token = null;
    try {
      token = await bearerToken(this.auth);
    } catch (e) {
      token = null;
      throw e;
    }
  
    const requestConfig = {
      url: `https://api.twitter.com/1.1/account_activity/all/${this.env}/webhooks.json`,
      options: {
        bearer: token,
      },
    };
  
    const response = await get(requestConfig);
    const error = tryError(
      response,
      (response) => new URIError(response, [
        `Cannot get webhooks. Please check that '${this.env}' is a valid environment defined in your`,
        'Developer dashboard at https://developer.twitter.com/en/account/environments, and that',
        `your OAuth credentials are valid and can access '${this.env}'. (HTTP status: ${response.statusCode})`].join(' ')));
  
    if (error) {
      throw error;
    }
  
    return response.body;
  }

  async removeWebhook(webhook) {
    await deleteWebhooks([webhook], this.auth, this.env);
  }

  async removeWebhooks() {
    const webhooks = await this.getWebhooks(this.auth, this.env);
    await deleteWebhooks(webhooks, this.auth, this.env);
  }
  
  async setNGrokWebhook() {
    if (this.ngrokSecret) {
      await ngrok.authtoken(this.ngrokSecret);
    }
    const url = await ngrok.connect(this.port);
    try {
      await this.setWebhook(`${url}${this.webhookPath}`);
      console.log('Webhook created.');
    } catch (e) {
      console.log('Cannot create webhook:', e);
      throw e;
    }
  }

  async start(webhookUrl = null) {
    
    if (!webhookUrl) {
      this.startServer();
      if (this.ngrokSecret) {
        await ngrok.authtoken(this.ngrokSecret);
      }
      const url = await ngrok.connect(this.port);
      webhookUrl = `${url}${WEBHOOK_ROUTE}`;      
    }
    
    try {
      await this.setWebhook(webhookUrl);
      console.log('Webhook created.');
    } catch (e) {
      console.log('Cannot create webhook:', e);
      throw e;
    }    
  }

  async subscribe({oauth_token, oauth_token_secret, screen_name = null}) {
    const auth = {
      consumer_key: this.auth.consumer_key,
      consumer_secret: this.auth.consumer_secret,
      token: oauth_token.trim(),
      token_secret: oauth_token_secret.trim(),
    };

    try {
      screen_name = screen_name || await verifyCredentials(auth);
    } catch (e) {
      screen_name = null;
      throw e;
    }

    const {subscriptions_count, provisioned_count} = await getSubscriptionsCount(auth);

    if (subscriptions_count === provisioned_count) {
      throw new TooManySubscriptionsError([`Cannot subscribe to ${screen_name}'s activities:`,
       'you exceeded the number of subscriptions available to you.',
       'Please remove a subscription or upgrade your premium access at',
       'https://developer.twitter.com/apps.',
       ].join(' '));
    }

    const requestConfig = {
      url: `https://api.twitter.com/1.1/account_activity/all/${this.env}/subscriptions.json`,
      options: {
        oauth: auth,      
      },
    };

    const response = await post(requestConfig);
    const error = tryError(
      response,
      (response) => new UserSubscriptionError(response));
    
    if (error) {
      throw error;
    }

    console.log(`Subscribed to ${screen_name}'s activities.`);
    updateSubscriptionCount(1);
    return true;  
  }

  async unsubscribe(userId) {
    const token = await bearerToken(this.auth);
    const requestConfig = {
      url: `https://api.twitter.com/1.1/account_activity/all/${this.env}/subscriptions/${userId}.json`,
      options: {
        bearer: token
      },
    };

    const response = await del(requestConfig);
    const error = tryError(
      response,
      (response) => new UserSubscriptionError(response));

      if (error) {
      throw error;
    }

    console.log(`Unsubscribed from ${userId}'s activities.`);
    updateSubscriptionCount(-1);
    return true;
  }
}

module.exports = {
  Autohook, 
  WebhookURIError, 
  UserSubscriptionError, 
  TooManySubscriptionsError, 
  validateWebhook,
  validateSignature,
};
